### DISCUSSION TITLE
GCC doesn't allow members in anonymous aggregate: struct Matrix4x4

### DISCUSSION THREAD

**Discussion by zwillikon** [leap71/PicoGK#47] [Engineering Discussion]
I struggle at compiling PicoGK with GCC, since the GNU compiler [doesn't allow members with constructors in anonymous aggregates ](https://stackoverflow.com/questions/41031424/constructor-not-allowed-in-anonymous-aggregate-string-in-struct)

I'm using (Linux) gcc 14.2.1, but older versions are also afflicted

```
struct Matrix4x4
{
    union
    {
        struct
        {
            Vector4 vec1;
            Vector4 vec2;
            Vector4 vec3;
            Vector4 vec4;
        };
        
        float   m[16];
    };
```

```
[ 94%] Building CXX object CMakeFiles/picogk.dir/Source/PicoGKGLViewer.cpp.o
In file included from ./PicoGKRuntime/Source/PicoGKGLViewer.h:39,
                 from ./PicoGKRuntime/Source/PicoGKGLViewer.cpp:37:
./PicoGKRuntime/Source/PicoGKTypes.h:278:21: error: member 'PicoGK::Vector4 PicoGK::Matrix4x4::<unnamed union>::<unnamed struct>::vec1' with constructor not allowed in anonymous aggregate
  278 |             Vector4 vec1;
      |                     ^~~~
./PicoGKRuntime/Source/PicoGKTypes.h:279:21: error: member 'PicoGK::Vector4 PicoGK::Matrix4x4::<unnamed union>::<unnamed struct>::vec2' with constructor not allowed in anonymous aggregate
  279 |             Vector4 vec2;
      |                     ^~~~
./PicoGKRuntime/Source/PicoGKTypes.h:280:21: error: member 'PicoGK::Vector4 PicoGK::Matrix4x4::<unnamed union>::<unnamed struct>::vec3' with constructor not allowed in anonymous aggregate
  280 |             Vector4 vec3;
      |                     ^~~~
./PicoGKRuntime/Source/PicoGKTypes.h:281:21: error: member 'PicoGK::Vector4 PicoGK::Matrix4x4::<unnamed union>::<unnamed struct>::vec4' with constructor not allowed in anonymous aggregate
  281 |             Vector4 vec4;
      |                     ^~~~
make[2]: *** [CMakeFiles/picogk.dir/build.make:90: CMakeFiles/picogk.dir/Source/PicoGKGLViewer.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:203: CMakeFiles/picogk.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

Just inventing names wasn't sufficient:
```
struct Matrix4x4
{
    union bar
    {
        struct foo
        {
            Vector4 vec1;
            Vector4 vec2;
            Vector4 vec3;
            Vector4 vec4;
        };
        
        float   m[16];
    };
```

leads to:

```
./Source/PicoGKTypes.h: In constructor 'PicoGK::Matrix4x4::Matrix4x4()':
./Source/PicoGKTypes.h:288:7: error: class 'PicoGK::Matrix4x4' does not have any field named 'vec1'
  288 |     : vec1(Vector4(1.0f, 0.0f, 0.0f, 0.0f)),
[..]
```

**Comment by soigniert** [comment]
Hello zwillikon,
I can report, that I had the same issue some months ago, when my Windows PC was not available for me.
I spent a whole weekend in order to compile the library and gave up (PicoGK 1.3, Ubuntu 22.04.4 LTS)
Multiple versions of _cmake_ with different _gcc_ and _clang_ (cmake 3.27.4, 3.29.2., 3.30.3 are the ones I can recall from the logs, I even set up a clean ubuntu, since I found that anaconda folder was always used to look for _clang++_ even when deleted, search path changed, aso...). In general I can say I progressed best with _clang_ instead _gcc_. With _clang_ I could at least compile runtime (1.1) from [CorrieVS](https://github.com/CorrieVS/PicoGKRuntime).
The problem with the specific code snipped I also ran into first place. It could be resolved somehow by writing it in a different form I cannot recall currently maybe like below, but I remember it only proceeded a bit with the compiling while running into another error. My latest explanation to myself was that the available cmake is just outdated. I would recommend to you, to ask at the fork of CorrieVS for help, since afaik leap71 uses macOS only and it's a concession to be provided for a windows runtime too. I can understand, since its not so easy with linux compatibility. Nonetheless I always hope for someone to keep the runtime available for linux users updated, since they lack in powerful design and CAx tools.
Best

```
struct Matrix4x4
{
    union
    {
        struct
        {
            Vector4 vectors[4];
        };
        
        float   m[16];
    } matrix;
        
    Matrix4x4()
    : matrix({{Vector4(1.0f, 0.0f, 0.0f, 0.0f),
              Vector4(0.0f, 1.0f, 0.0f, 0.0f),
              Vector4(0.0f, 0.0f, 1.0f, 0.0f),
              Vector4(0.0f, 0.0f, 0.0f, 1.0f)}})
    {

    }
```

**Comment by zwillikon** [comment]
Hi T Soignier,

thanks for your help. I wish'd I'd knew any C++ (just C)!

With old versions of clang I tend to get myriad errors against GCC's libstdc++ my because it is getting [ratcheted up by Mr. Wakely](https://github.com/gcc-mirror/gcc/commits/master/libstdc%2B%2B-v3) 

Your approach looks promising. Now only **m** is missing:

```
./Source/PicoGKTypes.h: In member function 'void PicoGK::Matrix4x4::operator*=(const PicoGK::Matrix4x4&)':
./Source/PicoGKTypes.h:304:28: error: 'm' was not declared in this scope
  304 |                     sum += m[row + i * 4] * other.m[col * 4 + i];
      |                            ^
./Source/PicoGKTypes.h:304:51: error: 'const struct PicoGK::Matrix4x4' has no member named 'm'
  304 |                     sum += m[row + i * 4] * other.m[col * 4 + i];
      |                                                   ^
./Source/PicoGKTypes.h:306:24: error: 'struct PicoGK::Matrix4x4' has no member named 'm'
  306 |                 result.m[col * 4 + row] = sum;
      |                        ^
```

**Reply by soigniert** [reply]
Hi @zwillikon , I got the latest PicoGK version finally running today using Cmake - Ubuntu 24.04.2 LTS; Ubuntu clang version 18.1.3 (1ubuntu1) Target: x86_64-pc-linux-gnu.
I want just to encourage you to also give it another shot with the latest version :)
Feel free to ask for help if you are stuck.

**Comment by LinKayser** [comment]
I tried it with gcc 15 on the 25.2 beta branch of the PicoGKRuntime, and it compiles fine. I think simply updating to gcc 15 does the trick.

I have no bandwidth to make an official LINUX version of PicoGK, but I am adding a Dockerfile to the repo that I will check regularly and compile against, to simply catch bugs that do not show up on Clang and MSVC.

**Reply by madscientist42** [reply]
It's all good on "official" support for right now.  Many of the hardcore Linux people (Waves hand...) are okay with no batteries included- just so long as it's not PAINFUL to try to get up and running against all the moving parts possible.  It may be some of us will convince you we're trustworthy enough to help you support it.  ;)

**Reply by madscientist42** [reply]
And to confirm, while it was clumsy/painful back a year ago, it's not any different now than trying to get C# up and running (That's where I had the initial, "fun," getting this going).  So, it meets with the ROE I threw out there- as long as the community can step up and help for when it breaks badly, I think the community will say, "well done," and help out.

Minor annoyances abound on getting the runtime built (We have, typically, your dependent runtimes, so doing Git Submodules is usually not the way to go.  It's more fragile from a CM perspective...but since you're doing this for Windows...nobody's going to do anything other than highlight a tighter rein on things...) but once you get past the, "that's now how we do things, usually..." it just builds and works.  I'm fighting with trying to make my first real Computational Geometry thing using RoverWheel, which is closest to my first application.  That's a mixed bag...but that's a differing discussion topic.

**Comment by LinKayser** [comment]
Here's the link to the instructions to set up Docker to compile PicoGK on LINUX: https://github.com/leap71/PicoGK/discussions/77
